### PR TITLE
plugins/wezterm: init

### PR DESCRIPTION
**Pull request by samos667**
This pull request introduces a new Neovim plugin for `wezterm` and integrates it into the existing plugin system. The changes include adding the `wezterm` plugin to the `molten` plugin list, defining the `wezterm` plugin in its own file, and adding test cases for the new plugin.

### New Plugin Addition:

* [`plugins/by-name/molten/default.nix`](diffhunk://#diff-230d660b843085745a52be69db8e4aa925f8c0297745d7b7bccbefc9c9ad2e29R83-R87): Added the `wezterm` plugin to the list of plugins and enabled it by default.
* [`plugins/by-name/wezterm/default.nix`](diffhunk://#diff-f8bd7524f04fe0ddf164336d51c3b20e34fcdc9c647a9948b32f068a4ffc3dd8R1-R21): Defined the `wezterm` plugin with its settings and maintainers.

### Testing:

* [`tests/test-sources/plugins/by-name/wezterm/default.nix`](diffhunk://#diff-62ec656d1f81a46b27b8ceb653a8e1db164bbc350bf6f34f9af9383d60b3c190R1-R12): Added test cases for the `wezterm` plugin to ensure it is enabled and its settings are correctly applied.

fixes https://github.com/nix-community/nixvim/issues/2596

### CODE REVIEW

File: tests/test-sources/plugins/by-name/wezterm/default.nix
Comment (khaneliman) [inline tests/test-sources/plugins/by-name/wezterm/default.nix:4]: Can you add a default test set with the default value for the provided option.

File: plugins/by-name/wezterm/default.nix
Comment (khaneliman) [inline plugins/by-name/wezterm/default.nix:14]:
```suggestion
      Whether to create plugin commands.
```
We don't need to specify the default in the description, the `true` from the arguments will generate that in the documentation.

File: plugins/by-name/wezterm/default.nix
Comment (MattSturgeon) [inline plugins/by-name/wezterm/default.nix:8]: > `error: vimPlugins.wezterm-nvim cannot be found in pkgs`

Looks like this [is in nixos-unstable now](https://search.nixos.org/packages?channel=unstable&show=vimPlugins.wezterm-nvim&from=0&size=50&sort=relevance&type=packages&query=wezterm-nvim). Rebasing on our latest `main` branch should bring a more up-to-date lock file and resolve this issue :upside_down_face:

File: plugins/by-name/molten/default.nix
Comment (MattSturgeon) [inline plugins/by-name/molten/default.nix:87]: This feels a little clunky. I think the indended wording is something like:

> ``If set to `"wezterm"`, `plugins.wezterm.enable` must be true.``

I'd also have a blankline (to trigger a paragraph break)

---

However it feels strange to have this notice for wezterm but not image.nvim? Maybe the original description was fine as it was?

Comment (samos667) [reply plugins/by-name/molten/default.nix:87]: idk because image.nvim directly refer to a plugin but not wezterm. But it's as you want, I don't really care

File: plugins/by-name/wezterm/default.nix
Comment (samos667) [reply plugins/by-name/wezterm/default.nix:8]: Should be good now

**Comment by MattSturgeon** [comment]
We still have `error: vimPlugins.wezterm-nvim cannot be found in pkgs`, so I guess we need to do another bump to the flake lock.

File: plugins/by-name/wezterm/default.nix
Comment (MattSturgeon) [reply plugins/by-name/wezterm/default.nix:8]: We still have `error: vimPlugins.wezterm-nvim cannot be found in pkgs`, so I guess we the most recent lock bump was slightly too old.

#2610 should resolve :crossed_fingers:

**Comment by khaneliman** [comment]
@samos667 We updated flake lock today so you should be able to rebase this, now.

**Review by GaetanLepage — CHANGES_REQUESTED** [review]

**Comment by samos667** [comment]
> error: attribute 'samos667' missing

Sadge :'(

**Comment by khaneliman** [comment]
> > error: attribute 'samos667' missing
> 
> 
> 
> Sadge :'(

If you're not in nixpkgs maintainers, you can add yourself to our maintainers file in another commit.

**Comment by MattSturgeon** [comment]
We also have a test failure:

```
defaults> ERROR: Error detected while processing /nix/store/XXX-init.lua:
defaults> Wezterm failed to find 'wezterm' executable
defaults> Wezterm failed to find 'wezterm' executable
```

Perhaps the nixpkgs package doesn't have wezterm marked as a dependency?

**Comment by MattSturgeon** [comment]
The following patch causes the test to pass, however I believe this should be fixed upstream in nixpkgs.

```diff
diff --git a/plugins/by-name/wezterm/default.nix b/plugins/by-name/wezterm/default.nix
index aecb332e..e95b36a4 100644
--- a/plugins/by-name/wezterm/default.nix
+++ b/plugins/by-name/wezterm/default.nix
@@ -1,4 +1,4 @@
-{ lib, ... }:
+{ lib, pkgs, ... }:
 let
   inherit (lib.nixvim) defaultNullOpts;
 in
@@ -9,6 +9,10 @@ lib.nixvim.neovim-plugin.mkNeovimPlugin {
 
   maintainers = [ lib.maintainers.samos667 ];
 
+  extraPackages = [
+    pkgs.wezterm
+  ];
+
   settingsOptions = {
     create_commands = defaultNullOpts.mkBool true ''
       Whether to create plugin commands.
```

Or if not\*, then we should have a nullable `weztermPackage` option.

```nix
weztermPackage = lib.mkPackageOption pkgs "wezterm" {
  nullable = true;
  # maybe also `default = null` and `example = [ "wezterm" ]`?
};
```

\* Perhaps nixpkgs decide that bundling the dep leads to too many potential issues, such as if users have an overridden/custom wezterm installed to their PATH?

**Comment by GaetanLepage** [comment]
> The following patch causes the test to pass, however I believe this should be fixed upstream in nixpkgs.
> 
> ```diff
> diff --git a/plugins/by-name/wezterm/default.nix b/plugins/by-name/wezterm/default.nix
> index aecb332e..e95b36a4 100644
> --- a/plugins/by-name/wezterm/default.nix
> +++ b/plugins/by-name/wezterm/default.nix
> @@ -1,4 +1,4 @@
> -{ lib, ... }:
> +{ lib, pkgs, ... }:
>  let
>    inherit (lib.nixvim) defaultNullOpts;
>  in
> @@ -9,6 +9,10 @@ lib.nixvim.neovim-plugin.mkNeovimPlugin {
>  
>    maintainers = [ lib.maintainers.samos667 ];
>  
> +  extraPackages = [
> +    pkgs.wezterm
> +  ];
> +
>    settingsOptions = {
>      create_commands = defaultNullOpts.mkBool true ''
>        Whether to create plugin commands.
> ```

This is still something we have not clearly agreed on in nixpkgs...
I think that the right approach here is having the suggested `weztermPackage` option.

File: tests/test-sources/plugins/by-name/wezterm/default.nix
Comment (GaetanLepage) [inline tests/test-sources/plugins/by-name/wezterm/default.nix:7]:
```suggestion
        weztermPackage = pkgs.wezterm;
```

File: tests/test-sources/plugins/by-name/wezterm/default.nix
Comment (GaetanLepage) [inline tests/test-sources/plugins/by-name/wezterm/default.nix:16]:
```suggestion
        weztermPackage = pkgs.wezterm;
```

**Review by GaetanLepage — CHANGES_REQUESTED** [review]

File: tests/test-sources/plugins/by-name/wezterm/default.nix
Comment (MattSturgeon) [reply tests/test-sources/plugins/by-name/wezterm/default.nix:7]: 🤣

File: plugins/by-name/wezterm/default.nix
Comment (MattSturgeon) [inline plugins/by-name/wezterm/default.nix:21]: What's the best default here? The package or null?

I'm assuming most users who enable this plugin will already have wezterm installed 🤔

**Comment by MattSturgeon** [comment]
@samos667 I've pushed up a couple tweaks. Let us know if you're happy and we should be good to approve & merge.

File: plugins/by-name/wezterm/default.nix
Comment (GaetanLepage) [reply plugins/by-name/wezterm/default.nix:21]: Usually we set the package.

File: plugins/by-name/molten/default.nix
Comment (GaetanLepage) [inline plugins/by-name/molten/default.nix:88]: Why not adding a warning to check this ? This is usually what we do.

**Comment by samos667** [comment]
Seems good to me !

File: plugins/by-name/wezterm/default.nix
Comment (MattSturgeon) [reply plugins/by-name/wezterm/default.nix:21]: This would achieve that:

```suggestion
    weztermPackage = lib.mkPackageOption pkgs "wezterm" {
      nullable = true;
    };
```

Is there any reason not to do that here?

**Comment by MattSturgeon** [comment]
I believe @GaetanLepage is hoping you'll change the following before we approve/merge:

1. Introduce a `warning` for https://github.com/nix-community/nixvim/pull/2597#discussion_r1876816570
2. Change the `weztermPackage` option to no longer default to null https://github.com/nix-community/nixvim/pull/2597#discussion_r1876736098

**Comment by khaneliman** [comment]
@samos667 Uninterested in continuing this?

**Comment by samos667** [comment]
Sorry, I can not invest myself as a contributor in this project, feel free to take the hand on this

**Pull request closed without merging**
